### PR TITLE
Upgrade to Go 1.21.1

### PR DESCRIPTION
**Pull request by braydonk**
## Description
This PR updates the Ops Agent to Go 1.21.1. It also adds a new export from the `internal/version` package to specify a Go version, which we will now use in a couple of places that we need to know the Go version.

## Related issue
[b/298656582](http://b/298656582)

## How has this been tested?
Tested as per instructions in the Go Upgrade dev docs, and now the integration tests in the PR will be the last step.

## Checklist:
- Unit tests
  - [x] Unit tests do not apply.
  - [ ] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [ ] Integration tests do not apply.
  - [x] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [x] This PR introduces no user visible changes.
  - [ ] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [x] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.
  - [ ] This PR bumps the version.

### CODE REVIEW

**Review by rafaelwestphal — APPROVED** [review]
lgtm

File: Dockerfile [hunk 1]
```diff
@@ -95,9 +95,9 @@ SHELL ["/bin/bash", "-c"]
 
 # Install golang
 ARG BUILDARCH
-ADD https://golang.org/dl/go1.20.2.linux-${BUILDARCH}.tar.gz /tmp/go1.20.2.tar.gz
+ADD https://go.dev/dl/go1.21.1.linux-${BUILDARCH}.tar.gz /tmp/go1.21.1.tar.gz
```
Comment (igorpeshansky) [inline Dockerfile:98]: Technically, the `golang.org` URL should still work, right? If the `go.dev` URLs are preferred now, do we want to decouple the version update from the URL update?

File: dockerfiles/compile.go [hunk 2]
```diff
@@ -46,6 +49,7 @@ func applyTemplate(template string, arguments templateArguments) string {
 		"{package_build}":     arguments.package_build,
 		"{tar_distro_name}":   arguments.tar_distro_name,
 		"{package_extension}": arguments.package_extension,
+		"{go_version}":        arguments.go_version,
```
Comment (igorpeshansky) [inline dockerfiles/compile.go:52]: Do we want to instead add an `ARG GO_VERSION=1.21.1` around line 27 in the `Dockerfile` to make future upgrade diffs smaller?

File: Dockerfile (hunk 1 above)
Comment (braydonk) [reply Dockerfile:98]: `golang.org` only exists now as a redirect, this shouldn't functionally change anything.

File: dockerfiles/compile.go (hunk 2 above)
Comment (braydonk) [reply dockerfiles/compile.go:52]: That's probably fine I can do that. Want to let this CI run finish first then will push that

**Comment by braydonk** [comment]
Both RHEL9 ARM failures and Windows Third Party App failure are flakes, so I am considering this a successful CI run. Going to address the last review comment and make sure the build passes again, then will merge.

**Comment by braydonk** [comment]
All the builds pass after the change, so I'm going to merge so this can get into the nightly

**Pull request merged**